Add verifier behaviour for conditional triples (#145)
ietf-rats-wg/draft-ietf-rats-corim: draft-ietf-rats-corim.md

[^issue]: Need to get agreement on how group membership is encoded. Tracked at https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/136

> *TODO: Can we find a better term than the placeholder "group"? If so then I think we should replace all mentions of "group" with the new term.*

#### Matching all Reference Values in a group against the Accepted Claims Set {#sec-match-all-ref-vals}

If a group contains at least one `reference-triple-record` and at least one
`endorsed-triple-record` then the Reference Values contained in the former
are used to decide whether to add Endorsed Values in the latter to the
Accepted Claims Set.

The Verifier SHALL try to match all Reference Values contained within all
`reference-triple-record`s within the group against Evidence and Endorsements
in the Accepted Claims Set. {{sec-match-one-ref-val}} describes how each
Reference Value is matched against the Accepted Claims Set.

If all Reference Values in a group match entries in the Accepted Claims Set
then all Endorsed Values from `endorsed-triple-record`s in the group are added
to the Accepted Claims Set (see {{sec-add-to-acs}}).

> If any Reference Value in a group does not match the Accepted Claims Set then

Reviewer: I think "Accepted Claims Set"  is ambiguous given the above statement that ACS may contain Endorsements as well as Evidence. The algorithm for accepting Evidence applies to Reference Values. If the Attester supplies no Evidence that is vetted by Reference Values, then no Endorsed values should be attributed IMO. Simple example, if Attester asserts it is an HSM, but no Ref Vals substantiate this, and a FIPS certifying body produces an Endorsement that says HSM is EAL4+, then the Attester has successfully achieved privilege escalation.

Author: We should probably discuss this more. My take is that Evidence is automatically added to the ACS with the authorization of the attester. Or more precisely with the authority key which authorized the attester (eg Intel DICE root if the verifier trusts that key).
For most relying parties, that evidence means nothing without a CoRIM to add a conditional endorsement. The FIPS certifying body could say that an entity with measurements A=A1 and B=B1 authorized by key K is FIPS certified.
As we discussed, I think authorized by deserves its own section.

Author: I will address this as part of the patch for issue #140

all `endorsed-triple-record`s in the group are silently ignored.

Each group is processed independently of other groups. If `reference-triple-record`s
in a group do not match the Accepted Claims Set then this does not affect
the processing of other groups.

#### Matching a Conditional Endorsement Triple against the Accepted Claims Set

A Conditional Endorsement Triple {{sec-comid-triple-cond-end}} is an efficient

> encoding of a Reference Value and an Endorsed Value.

Reviewer: ```suggestion
encoding of a Reference Value that describes the result of previously accepted Claims as the condition for accepting additional Endorsed Values.
```

Author: The next couple of lines say that, so I don't think it needs to say it here as well.

> The triple's `stateful-environment-record` field contains both the `environment-map` and `measurement-map` parts of the Reference Value.

Reviewer: ```suggestion
The triple's `stateful-environment-record` field constitutes the Reference Value which contains both `environment-map` and `measurement-map` Claims.
```

Author: I think a change similar to what you propose is good, but it depends on which name we use to replace "Reference Value"

> The Endorsed Value is encoded using the same `environment-map`, plus the

Reviewer: This looks incorrect. The triple object is `measurement-values-map` not an environment-map.

Author: We need to add the MVM to an EM somewhere. As I understand it, the MVM is added to the same EM as the one in the stateful-environment-record.

`measurement-value-map` field within the Conditional Endorsement Triple.

> For each Conditional Endorsement Triple, the verifier SHALL attempt to match the

Reviewer: ```suggestion
For each Conditional Endorsement Triple, the verifier SHALL match the
```

Author: agree

> Reference Value against entries in the Accepted Claims Set (see {{sec-match-one-ref-val}}).

Reviewer: ```suggestion
Reference Value, consisting of a `stateful-environment-record`, against entries in the Accepted Claims Set (see {{sec-match-one-ref-val}}).
```

If it matches then the verifier SHALL add the Endorsed value to the
Accepted Claims Set (see {{sec-add-to-acs}}).

>> *TODO: Is a Conditional Endorsement Triple affected by `reference-triple-record`s

> in the same group? I think not. Does this need to be mentioned?*

Reviewer: It can be if the reference-triple-record contains claims that are in the stateful-environment-record. The expectation is the reference-triples are processed before the conditional triples. (I think).

Author: Again, I am asking about whether CET is a group of its own or is part of another group

#### Matching a Conditional Endorsement Series Triple against the Accepted Claims Set

A Conditional Endorsement Series Triple {{sec-comid-triple-cond-series}} is an efficient

> encoding of a series of alternative Reference Values, each with a corresponding

Reviewer: I think this is misstated. The series claims are endorsed values, of which one in the series is selected for inclusion in the ACS. The RVs are not a "series" but a stateful environment.

Author: The is the introductory description of what the triple does, IMHO it should describe the rest of the section in informative words so that a reader knows what is coming next.
I used the word "series" because the verifier processes them in order.
I haven't found a non-circular definition of "stateful environment" so I don't think that is a good term to use instead. It might be if we define it appropriately. Perhaps what I called "Reference Value" is what you think of as "stateful environment"?

> Endorsed Value. The Verifier adds the Endorsed Values corresponding to the first

Reviewer: ```suggestion
Endorsed Value.
The Verifier adds the first Endorsed Value in the series that matches the stateful environment.
```

> Reference Value which matches.

Reviewer: ```suggestion
The stateful enviroment acts like a Reference Value which matches Claims contained in the Accepted Claims Set.
```

Author: I am wondering whether to do s/Reference Value/stateful environment/ or something like that throughout the whole diff

> All Reference Values and Endorsed Values in the triple use the same `environment-map`

Reviewer: I don't know what these two lines are trying to say, but it seems incorrect.

Author: I may be misunderstanding your expectations for CET

Reference Values are encoded in two parts of this triple.

Reference Values which apply to every Endorsed Value are encoded in the
`stateful-environment-record`, in the same way as for the
Conditional Endorsement Triple. If any of these Reference Values
do not match Evidence or Endorsements in the Accepted Claims Set then then the
entire triple is silently ignored.

Reference Values which apply to only one Endorsed Value are encoded, with their
matching Endorsed Value, in an array of `conditional-series-record`s.
Each `conditional-series-record` encodes the `measurement-value-map` part for its
Reference Values and for the corresponding Endorsed Value.

The entries in the `conditional-series-record` array are processed in order. The verifier
SHALL attempt to match the Reference Value part of each entry against entries
in the Accepted Claims Set (see {{sec-match-one-ref-val}}).
If the Reference Value matches then the verifier SHALL add the corresponding
Endorsed Value to the Accepted Claims Set (see {{sec-add-to-acs}}) and
SHALL skip processing the remaining entries in the array.

>> *TODO: Is a Conditional Endorsement Series Triple affected by

> `reference-triple-record`s in the same group? I think not.

Reviewer: It could be if the stateful-environment-record mentions endorsement claims that were added via a reference triple.

Author: I agree that the reference value part of Conditional Endorsement Series Triple could match against an endorsement added by an earlier group.
I guess what I was asking is whether CEST is in a group or not. If it's in a group then a mismatched reference value in that group could stop it from being processed. That seems odd to me, so I would rather say that CEST is not in a group, but I want to make sure everyone is happy with what I'm writing.

Does this need to be mentioned?*

#### Matching a Reference Value against the Accepted Claims Set {#sec-match-one-ref-val}

> This section describes how a Reference Value is matched against Evidence or Endorsements

Reviewer: What are some use cases for Ref Values matching Endorsements?

Author: The main use case is hierarchical CoRIMs. 
- CoMID A has reference values which are expected to match against measurements A and B, if they match it adds endorsement E. 
- CoMID C contains reference values which are expected to match against measurements C and D, and another which is expected to match against endorsement E if authorized by A. If all match then it adds endorsement F.
- The relying party is looking for endorsement F.
I'll put this in a separate PR

in the Accepted Claims Set.
If any part of the processing indicates that the Reference Value does not match
then the remaining steps in this section are skipped for that group.

> A Reference Value consists of an `environment-map` plus a `measurement-map`. It is

Reviewer: I still think this is too narrowly stated. If taken literally, the identity-triples, attest-key-triples, conditional triples and coswid-triples have to be discarded because their "measurements" aren't describes as `measurement-map`.

Author: As you see below I think the spec needs a name for `environment-map` plus `measurement-map` when it appears in conditional triples as well as when it appears in reference-values-triple-record.
I am quite happy to use a new name for EM+MM.
What should it be?

Author: Possible suggestions:
environment-measurement-pair
claim tuple

Reviewer: > environment-measurement-pair

Fine! but probably would not use hyphens with pair. E.g.: "... an environment-measurement pair..." or "... an environment-measurement pairing..."
